XenAdmin/Commands/ManageCPUGroupCommand.cs

```diff
@@ -0,0 +1,113 @@
+ *     following disclaimer. 
+ * *   Redistributions in binary form must reproduce the above 
+ *     copyright notice, this list of conditions and the 
+ *     following disclaimer in the documentation and/or other 
+ *     materials provided with the distribution. 
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND 
+ * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+ * SUCH DAMAGE.
+ */
+
+using System.Collections.Generic;
+using System.Windows.Forms;
+using XenAdmin.Network;
+using XenAPI;
+using XenAdmin.Core;
+using XenAdmin.Dialogs;
+
+namespace XenAdmin.Commands
+{
+    /// <summary>
+    /// Launches the Manage CPU Group dialog.
+    /// </summary>
+    internal class ManageCPUGroupCommand : Command
+    {
+        /// <summary>
+        /// Initializes a new instance of this Command. The parameter-less constructor is required if 
+        /// this Command is to be attached to a ToolStrip menu item or button. It should not be used in any other scenario.
+        /// </summary>
+        public ManageCPUGroupCommand()
+        {
+        }
+
+        public ManageCPUGroupCommand(IMainWindow mainWindow, IEnumerable<SelectedItem> selection)
+            : base(mainWindow, selection)
+        {
+        }
+
+        public ManageCPUGroupCommand(IMainWindow mainWindow, IXenConnection connection)
+            : base(mainWindow, Helpers.GetPoolOfOne(connection))
+        {
+        }
+
+        public ManageCPUGroupCommand(IMainWindow mainWindow, Pool pool)
+            : base(mainWindow, pool)
+        {
+        }
+
+        public ManageCPUGroupCommand(IMainWindow mainWindow, Host host)
+            : base(mainWindow, host)
+        {
+        }
+
+        protected override bool CanExecuteCore(SelectedItemCollection selection)
+        {
+            // this check ensures there's no cross-pool and items don't contain VM
+            if (selection.FirstAsXenObject != null && selection.FirstAsXenObject.Connection != null && selection.FirstAsXenObject.Connection.IsConnected &&
+               (selection.PoolAncestor != null || selection.HostAncestor != null) && !itemsContainVM(selection))
+                return true;
+
+            return false;
+        }
+
+
+        private bool itemsContainVM(SelectedItemCollection selection)

```

Comment:
No need to implement a new method to check we have a VM. You can use one of the ready predicates in SelectedItemCollection.cs, namely ContainsOneItemOfType<VM>.

Reply:
In my opinion, the method named ContainsOneItemOfType predicates only one item of specified type. However, I don't know how the method works when user choose more than one item contain hosts and VMs.

Reply:
In design of REQ-155,  the GUI won't all user to use the "Manage CPU groups..." button if user select a VM. There is a case that user select more than one item contain VM. In that case, we shouldn't allow user to use the button.

Reply:
Ahh indeed, you're right, ContainsOneItemOfType is only for 1 item, my bad! In this case it would be better to maker a new reusable function in that class, AtLeastOneXenObjectIs.
However, Mihaela is right, checking the pool or host ancestor should be enough, and it's not wrong if the item is enabled when a VM in the pool is selected. There are other pool or host items on the main menu that are enabled if a VM with that pool or host ancestor is selected.
One more thing I forgot to mention yesterday is we should probably add the item in the Pool's right click menu. For this you'll need to add the item in the SinglePool class in xenadmin\XenAdmin\Commands\Controls\ContextMenuBuilder.cs

Reply:
Thank you very much. I will remove it.